p1-staking-pool-dyn/src/lib.rs

```diff
@@ -275,52 +261,42 @@ impl Contract {
     }
 
     #[private]
-    pub fn mint_callback(&mut self, user: AccountId, amount: U128, close: bool) {
-        env_log!(
-            "mint_callback, env::promise_results_count()={}",
-            env::promise_results_count()
-        );
+    pub fn mint_callback(&mut self, user: AccountId, amount: U128) {
         // after the async call to mint rewards for the user
         assert_eq!(
             env::promise_results_count(),
             1,
             "{}",
             ERR25_WITHDRAW_CALLBACK
         );
+        // check prev promise result
         match env::promise_result(0) {
             PromiseResult::NotReady => unreachable!(),
 
             PromiseResult::Successful(_) => {
                 env_log!("cheddar rewards withdrew {}", amount.0);
                 self.total_rewards += amount.0;
-                if close {
-                    // AUDIT: TODO: Should check that the vault doesn't have `.staked > 0`, because
-                    //    in case of weird async race conditions, the vault might get new staked balance.
-                    self.vaults.remove(&user);
-                    env::log(b"account closed");
-                }
             }
 
             PromiseResult::Failed => {
                 // mint failed, restore cheddar rewards
-                // AUDIT: TODO: If the vault was closed before by another TX, then the contract
-                //     should recover the vault here.
-                let mut vault = self.vaults.get(&user).expect(ERR10_NO_ACCOUNT);
-                // AUDIT: TODO: You may also want to ping the vault.
-                vault.rewards = amount.0;
-                self.vaults.insert(&user, &vault);
+                // recover the vault
+                // AUDIT: TODO?: You may also want to ping the vault.
+                let mut vault = self.get_vault_or_default(&user);
+                vault.rewards += amount.0;
+                self.save_vault(&user, &vault);
             }
         }
     }
 
     #[private]
     pub fn mint_callback_finally(&mut self, user: &AccountId) {
         //Check if rewards were withdrew
-        if let Some(vault) = self.vaults.get(&user) {
-            if vault.rewards != 0 {
-                //if there are cheddar rewards, means the cheddar transfer failed
-                panic!("{}", "cheddar transfer failed");
-            }
+        // check the vault
+        let vault = self.get_vault_or_default(&user);
+        if vault.rewards != 0 {
+            //if there are cheddar rewards, means the cheddar transfer failed
+            panic!("{}", "cheddar transfer failed");
```

Comment:
This can be wrong because this function is happening at least 3 blocks later than the first call.
+ `withdraw_crop` doesn't close an account and if user pinged the contract in the meantime, the `vault.rewards` won't be 0.
+ in a very rare case, user can close an account and stake again - so `vault.rewards` will be non zero.

Reply:
very rare case, almost impossible. The user can retry the tx and withdraw again.